### PR TITLE
[Optimization] Calculate global parameter and local parameter at master.

### LINKED ISSUE #10703: [Optimization] Global parameter and local parameter calculation external expansion.
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar feature requirement.

### Description

Global parameter and local parameter calculation external expansion.

### Use case

Create a workflow with global and local parameters.

### Related issues

No.

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by WangJPLeo**
## Purpose of the pull request

1. Add an external extension class to the service.
2. The global parameters and local parameters are calculated and prepared in the master service.
3. The local parameters in the worker service will be used directly.

## Brief change log

close #10703 

## Verify this pull request

Manually verified the change by testing locally.

### CODE REVIEW

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/expand/DolphinSchedulerCuringGlobalParams.java [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.apache.dolphinscheduler.plugin.task.api.TaskConstants.PARAMETER_TASK_EXECUTE_PATH;
+import static org.apache.dolphinscheduler.plugin.task.api.TaskConstants.PARAMETER_TASK_INSTANCE_ID;
+
+@Component
+public class DolphinSchedulerCuringGlobalParams implements CuringParamsService {
```
Comment (caishunfeng) [inline dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/expand/DolphinSchedulerCuringGlobalParams.java]:
```suggestion
public class CuringGlobalParams implements CuringParamsService {
```

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/expand/DolphinSchedulerCuringGlobalParams.java [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+
+        if (timeParams != null) {
+            allParamMap.putAll(timeParams);
+        }
+        allParamMap.putAll(globalMap);
+        Set<Map.Entry<String, String>> entries = allParamMap.entrySet();
+        Map<String, String> resolveMap = new HashMap<>();
+        for (Map.Entry<String, String> entry : entries) {
+            String val = entry.getValue();
+            if (val.startsWith("$")) {
```
Comment (caishunfeng) [inline dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/expand/DolphinSchedulerCuringGlobalParams.java]: use constant

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/expand/DolphinSchedulerCuringGlobalParams.java [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+    }
+
+    @Override
+    public Map<String, Property> paramParsingPreparation(TaskExecutionContext taskExecutionContext, AbstractParameters parameters, ProcessInstance processInstance) {
+        Preconditions.checkNotNull(taskExecutionContext);
+        Preconditions.checkNotNull(parameters);
+        setGlobalParamsMap(taskExecutionContext);
+        Map<String, Property> globalParams = ParamUtils.getUserDefParamsMap(taskExecutionContext.getDefinedParams());
+        Map<String,String> globalParamsMap = taskExecutionContext.getDefinedParams();
+        org.apache.dolphinscheduler.spi.enums.CommandType commandType = org.apache.dolphinscheduler.spi.enums.CommandType.of(taskExecutionContext.getCmdTypeIfComplement());
```
Comment (caishunfeng) [inline dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/expand/DolphinSchedulerCuringGlobalParams.java]:
```suggestion
        CommandType commandType = CommandType.of(taskExecutionContext.getCmdTypeIfComplement());
```

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/expand/DolphinSchedulerCuringGlobalParams.java [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+        if (localParams.size() != 0) {
+            globalParams.putAll(localParams);
+        }
+
+        Iterator<Map.Entry<String, Property>> iter = globalParams.entrySet().iterator();
+        while (iter.hasNext()) {
+            Map.Entry<String, Property> en = iter.next();
+            Property property = en.getValue();
+
+            if (StringUtils.isNotEmpty(property.getValue()) && property.getValue().startsWith("$")) {
```
Comment (caishunfeng) [inline dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/expand/DolphinSchedulerCuringGlobalParams.java]: same here

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/expand/DolphinSchedulerCuringGlobalParams.java [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+                 *  note: the global parameters of the process instance here are solidified parameters,
+                 *  and there are no variables in them.
+                 */
+                String val = property.getValue();
+                // whether external scaling calculation is required
+                if (timeFunctionNeedExpand(val)) {
+                    Map<String, String> cmdParam = JSONUtils.toMap(processInstance.getCommandParam());
+                    val = timeFunctionExtension(taskExecutionContext.getProcessInstanceId(), cmdParam.get(Constants.SCHEDULE_TIMEZONE), val);
+                } else {
+                    val  = org.apache.dolphinscheduler.plugin.task.api.parser.ParameterUtils.convertParameterPlaceholders(val, params);
```
Comment (caishunfeng) [inline dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/expand/DolphinSchedulerCuringGlobalParams.java]:
```suggestion
                    val  = ParameterUtils.convertParameterPlaceholders(val, params);
```

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/expand/DolphinSchedulerCuringGlobalParams.java [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+                }
+                property.setValue(val);
+            }
+        }
+        if (org.apache.dolphinscheduler.plugin.task.api.utils.MapUtils.isEmpty(globalParams)) {
```
Comment (caishunfeng) [inline dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/expand/DolphinSchedulerCuringGlobalParams.java]:
```suggestion
        if (MapUtils.isEmpty(globalParams)) {
```

File: dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/parser/ParamUtils.java [hunk 7]
```diff
@@ -43,77 +33,81 @@ public class ParamUtils {
...
-                String val = property.getValue();
-
-                val  = ParameterUtils.convertParameterPlaceholders(val, params);
-                property.setValue(val);
-            }
-        }
-
-        return globalParams;
-    }
+//    public static Map<String, Property> convert(TaskExecutionContext taskExecutionContext, AbstractParameters parameters) {
```
Comment (caishunfeng) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/parser/ParamUtils.java]: remove if no use.

File: dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/parser/ParameterUtils.java [hunk 8]
```diff
@@ -155,54 +151,59 @@ public static void setInParameter(int index, PreparedStatement stmt, DataType da
...
-
-        for (Property property : globalParamList) {
-            String val = globalMap.get(property.getProp());
-            if (val != null) {
-                property.setValue(val);
-            }
-        }
-        return JSONUtils.toJsonString(globalParamList);
-    }
+//    public static String curingGlobalParams(Map<String, String> globalParamMap, List<Property> globalParamList,
```
Comment (caishunfeng) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/parser/ParameterUtils.java]: same here.

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/expand/DolphinSchedulerCuringGlobalParams.java [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+            String val = globalMap.get(property.getProp());
+            if (val != null) {
+                property.setValue(val);
+            }
+        }
+        return JSONUtils.toJsonString(globalParamList);
+    }
+
+    @Override
+    public Map<String, Property> paramParsingPreparation(TaskExecutionContext taskExecutionContext, AbstractParameters parameters, ProcessInstance processInstance) {
```
Comment (caishunfeng) [inline dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/expand/DolphinSchedulerCuringGlobalParams.java]: please add some comments.

File: dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/parser/ParamUtils.java (hunk 7 above)
Comment (lenboo) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/parser/ParamUtils.java]: You can delete unused code.

**Review by lenboo — CHANGES_REQUESTED** [review]

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/expand/DolphinSchedulerCuringGlobalParams.java (hunk 1 above)
Comment (WangJPLeo) [reply dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/expand/DolphinSchedulerCuringGlobalParams.java]: ok.

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/expand/DolphinSchedulerCuringGlobalParams.java (hunk 2 above)
Comment (WangJPLeo) [reply dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/expand/DolphinSchedulerCuringGlobalParams.java]: ok.

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/expand/DolphinSchedulerCuringGlobalParams.java (hunk 3 above)
Comment (WangJPLeo) [reply dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/expand/DolphinSchedulerCuringGlobalParams.java]: Proven to be replaceable.

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/expand/DolphinSchedulerCuringGlobalParams.java (hunk 9 above)
Comment (ruanwenjun) [reply dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/expand/DolphinSchedulerCuringGlobalParams.java]: I hope we can add comments and example in the interface, unless the interface is easy to know.

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/builder/TaskExecutionContextBuilder.java [hunk 10]
```diff
@@ -131,6 +136,13 @@ public TaskExecutionContextBuilder buildK8sTaskRelatedInfo(K8sTaskExecutionConte
         taskExecutionContext.setK8sTaskExecutionContext(k8sTaskExecutionContext);
         return this;
     }
+
+    public TaskExecutionContextBuilder buildParamInfo(CuringParamsService curingParamsService, AbstractParameters baseParam, ProcessInstance processInstance) {
+        Map<String, Property> propertyMap = curingParamsService.paramParsingPreparation(taskExecutionContext, baseParam, processInstance);
+        taskExecutionContext.setPrepareParamsMap(propertyMap);
+        return this;
+    }
```
Comment (ruanwenjun) [inline dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/builder/TaskExecutionContextBuilder.java]: It's better to calculate the param in the upper layer.
```suggestion
    public TaskExecutionContextBuilder buildParamInfo(Map<String, Property> propertyMap) {
        taskExecutionContext.setPrepareParamsMap(propertyMap);
        return this;
    }
```

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteRunnable.java [hunk 11]
```diff
@@ -255,7 +254,7 @@ public WorkflowExecuteRunnable(ProcessInstance processInstance
             , ProcessAlertManager processAlertManager
             , MasterConfig masterConfig
             , StateWheelExecuteThread stateWheelExecuteThread
-            , CuringGlobalParamsService curingGlobalParamsService) {
+            , CuringParamsService curingGlobalParamsService) {
```
Comment (ruanwenjun) [inline dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkflowExecuteRunnable.java]:
```suggestion
            , CuringParamsService curingParamsService) {
```

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/expand/DolphinSchedulerCuringGlobalParams.java [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+        Preconditions.checkNotNull(taskExecutionContext);
+        Preconditions.checkNotNull(parameters);
```
Comment (ruanwenjun) [inline dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/expand/DolphinSchedulerCuringGlobalParams.java]: It's better to use `@NonNull` in the method to replace Preconditions.checkNotNull.

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/expand/DolphinSchedulerCuringGlobalParams.java [hunk 13]
```diff
@@ -0,0 +1,199 @@
...
+        setGlobalParamsMap(taskExecutionContext);
```
Comment (ruanwenjun) [inline dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/expand/DolphinSchedulerCuringGlobalParams.java]: So this method will affect `taskExecutionContext`, I think this should be note in the comment.

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/expand/DolphinSchedulerCuringGlobalParams.java [hunk 14]
```diff
@@ -0,0 +1,199 @@
...
+        org.apache.dolphinscheduler.spi.enums.CommandType commandType = org.apache.dolphinscheduler.spi.enums.CommandType.of(taskExecutionContext.getCmdTypeIfComplement());
+        Date scheduleTime = taskExecutionContext.getScheduleTime();
+
+        // combining local and global parameters
+        Map<String, Property> localParams = parameters.getInputLocalParametersMap();
+
+        //stream pass params
+        Map<String, Property> varParams = parameters.getVarPoolMap();
+
+        if (globalParams.size() == 0 && localParams.size() == 0 && varParams.size() == 0) {
```
Comment (ruanwenjun) [inline dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/expand/DolphinSchedulerCuringGlobalParams.java]: Use `isEmpty()` is better than `.size() == 0`

**Review by ruanwenjun — COMMENTED** [review]
I don't see the whole, just add some comments about code style.

**Review by ruanwenjun — APPROVED** [review]
LGTM

**Comment by lenboo** [comment]
+1

**Review by lenboo — APPROVED** [review]
+1

**Pull request merged**